packages/x-data-grid-pro/src/hooks/features/serverSideLazyLoader/useGridIntersectionObserver.tsx

```diff
@@ -0,0 +1,128 @@
+import * as React from 'react';
+import { RefObject } from '@mui/x-internals/types';
+import { useGridSelector, useGridApiMethod, gridDimensionsSelector } from '@mui/x-data-grid';
+import {
+  GridInfiniteLoaderPrivateApi,
+  useTimeout,
+  gridHorizontalScrollbarHeightSelector,
+} from '@mui/x-data-grid/internals';
+import useEventCallback from '@mui/utils/useEventCallback';
+import { styled } from '@mui/system';
+import { GridPrivateApiPro } from '../../../models/gridApiPro';
+import { DataGridProProcessedProps } from '../../../models/dataGridProProps';
+
+const InfiniteLoadingTriggerElement = styled('div')({
+  position: 'sticky',
+  left: 0,
+  width: 0,
+  height: 0,
+});
+
+/**
+ * @requires useGridDimensions (method) - can be after
+ */
+export const useGridIntersectionObserver = (
+  apiRef: RefObject<GridPrivateApiPro>,
+  props: Pick<
+    DataGridProProcessedProps,
+    'onRowsScrollEnd' | 'dataSource' | 'lazyLoading' | 'rowsLoadingMode' | 'scrollEndThreshold'
+  >,
+): void => {
+  const isReady = useGridSelector(apiRef, gridDimensionsSelector).isReady;
+  const observer = React.useRef<IntersectionObserver>(null);
+  const updateTargetTimeout = useTimeout();
+  const triggerElement = React.useRef<HTMLElement | null>(null);
+
+  const isEnabledClientSide = props.rowsLoadingMode === 'client' && !!props.onRowsScrollEnd;
+  const isEnabledServerSide = props.dataSource && props.lazyLoading;
```

Comment:
Would it make sense to skip running the intersection observer logic for viewport loading mode?
It'd also remove the need to check it explicitly in `handleIntersection()`
```suggestion
  const isEnabledServerSide = props.dataSource && props.lazyLoading && (rowCount !== undefined || rowCount !== -1);
```

Reply:
I would rather not rely on this, because the condition can change in the other hook.

Reply:
Fine, not a big deal anyways.
If we ever make `loadingTrigger` part of state, we could use it here to avoid extra computation.